### ISSUE TITLE
TypeError: Cannot read property 'resurrectTimeout' of undefined at ConnectionPool.resurrect

### ISSUE THREAD

**Issue by kokozhang** [elastic/elasticsearch-js#940]
## 🐛 Bug Report
Hello. When I called esClint.index() I got: 
`this.connections.get(this.dead[0])` is undefined at ConnectionPool.js

## To Reproduce

```js
const elasticsearch = require('@elastic/elasticsearch');

const esClient = new elasticsearch.Client({ node: ESHost.split(',').map(host => host.startsWith('http://') ? host : `http://${host}`), sniffOnStart: true });

esClient.index(q)
```

## Expected behavior
Index this document. And Code compiles without error

## Actual behavior
```js
TypeError: Cannot read property 'resurrectTimeout' of undefined
    at ConnectionPool.resurrect (/var/task/node_modules/@elastic/elasticsearch/lib/ConnectionPool.js:124:47)
    at Transport.getConnection (/var/task/node_modules/@elastic/elasticsearch/lib/Transport.js:332:25)
    at makeRequest (/var/task/node_modules/@elastic/elasticsearch/lib/Transport.js:124:30)
    at Transport.request (/var/task/node_modules/@elastic/elasticsearch/lib/Transport.js:316:15)
    at Promise (/var/task/node_modules/@elastic/elasticsearch/lib/Transport.js:90:14)
    at new WrappedPromise (/var/task/node_modules/async-listener/es6-wrapped-promise.js:13:18)
    at Transport.request (/var/task/node_modules/@elastic/elasticsearch/lib/Transport.js:89:14)
    at _index (/var/task/node_modules/@elastic/elasticsearch/api/api/index.js:148:12)
    at Client._lazyLoad [as index] 
```

## Your Environment

- *node version*: 8
- `@elastic/elasticsearch` *version*: 6.8.1
- *os*: Linux
- *Deployed in Lambda*

**Comment by delvedor** [comment]
Hello!
I'm not able to reproduce the issue on my local machine, can you that with https://github.com/delvedor/es-reproduce-issue?

**Comment by ganxiyun** [comment]
How to reproduce

1. ESHost list is provided in initialization
2. sniffOnStart is true
3. one connection returns error response so the connection is marked as dead before the finish of `update (connections)`

**Comment by delvedor** [comment]
Are you able to reproduce it also with the latest release of the client?
It looks like that the issue is caused by the "old" implementation of the connection pool (previous to https://github.com/elastic/elasticsearch-js/pull/913) where the `dead` queue was not cleaned up after an update.

The current implementation, which has been released in `6.8.2` and `7.3.0`, should fix your problem. Are you able to reproduce the issue even with the latest version?
If so, can you use https://github.com/delvedor/es-reproduce-issue to show how to reproduce the issue?
Thanks!

**Comment by JerseyGood** [comment]
@delvedor Even in latest code `6.8.2` this issue stands.

Here's what happened.
1. Initialize client with node IP and `sniffOnStart:true`
2. The first sniff request was sent 
3. A search query request was sent with node IP as connection.id 
4. The response of the first sniff request come back and update ConnectionPool to use UUID instead of Node IP as connection id.
5. Somehow the search query failed and markDead the connection with Node IP. (Don't know why yet)
6. Next search query will try to resurrect this connection but the Connection.ID in dead list is Node IP which cannot be found in ConnectionPool.

This happened quite often in Lambda since the client initialization was triggered (and the container startup) by the query request.

**Comment by JerseyGood** [comment]
@delvedor 
The new code did clean up the `dead` list but isn't this code
https://github.com/elastic/elasticsearch-js/blob/a4bff1c2f0a870781ff9171e351530f98621142e/lib/pool/ConnectionPool.js#L218
but what if you got 2 dead connections that are not tracked by `this.connections`?
Wouldn't the 2nd `splice` cause trouble?

**Comment by ganxiyun** [comment]
> Are you able to reproduce it also with the latest release of the client?
> It looks like that the issue is caused by the "old" implementation of the connection pool (previous to #913) where the `dead` queue was not cleaned up after an update.
> 
> The current implementation, which has been released in `6.8.2` and `7.3.0`, should fix your problem. Are you able to reproduce the issue even with the latest version?
> If so, can you use https://github.com/delvedor/es-reproduce-issue to show how to reproduce the issue?
> Thanks!

I have created test cases, could you please have a check?

**Comment by vermadev95** [comment]
Is this issue resolved? I found this on the latest release on ES 7.9.1
